File: src/erl/dating.erl
-module(dating).
-export([start/0]).

calculateAge(Birthday) ->
    {BirthdayYear, BirthdayMonth, BirthdayDay} = Birthday,
    {{NowYear, NowMonth, NowDay}, _} = calendar:local_time(),
    Age = NowYear - BirthdayYear,
    if 
        NowMonth > BirthdayMonth -> Age;
        NowMonth < BirthdayMonth -> Age -1;
        true -> if
                NowDay >= BirthdayDay -> Age;
                true -> Age + 1 
            end
    end.

parseDate(String) ->
    [Year, Month, Day] = string:tokens(String, "-"),
    {list_to_integer(Year), list_to_integer(Month), list_to_integer(Day)}.

parseTime(String) ->
     [Hours, Minutes, Seconds] = string:tokens(String, ":"),
     {list_to_integer(Hours), list_to_integer(Minutes), list_to_integer(Seconds)}.

parseDateTime(String) ->
    [Date, Time] = string:tokens(String, "T"),
    {parseDate(Date), parseTime(Time)}.

start() -> 
    Birthday = {1939, 10, 27},
    {BirthdayYear, BirthdayMonth, BirthdayDay} = Birthday,
    io:fwrite("John Cleese was born in ~w-~2..0w-~2..0w and is ~w years old.~n", [BirthdayYear, BirthdayMonth, BirthdayDay, calculateAge(Birthday)]),

    {{Year, Month, Day}, { Hours, Minutes, Seconds}} = parseDateTime("1975-11-10T01:25:00"),
    io:fwrite("Parsed date is ~w-~2..0w-~2..0wT~2..0w:~2..0w:~2..0w", [Year, Month, Day, Hours, Minutes, Seconds]),

    erlang:halt().

File: src/cpp/dating.cpp
#include <iostream>
#include <ctime>
#include <cstdio>

using namespace std;

int getAgeInYears(tm birthday)
{
    time_t now = time(NULL);
    tm localNow = *localtime(&now);

    return localNow.tm_year - birthday.tm_year - (localNow.tm_yday < birthday.tm_yday ? 1 : 0);
}

tm parseDate(char* value)
{
    tm result;
    sscanf(value, 
        "%i-%i-%iT%i:%i:%i", 
        &(result.tm_year), 
        &(result.tm_mon), 
        &(result.tm_mday), 
        &(result.tm_hour), 
        &(result.tm_min), 
        &(result.tm_sec));

    result.tm_year -= 1900;
    result.tm_mon -= 1;

    return result;
}

int main()
{
    tm birthday; 
    birthday.tm_year = 1939 - 1900;
    birthday.tm_mon = 10 - 1;
    birthday.tm_mday = 27;

    char formattedBirthday[50];
    strftime(formattedBirthday, 50, "%F", &birthday);
    int age = getAgeInYears(birthday);
    cout << "John Cleese was born in " << formattedBirthday << " and is " << age << " years old." << endl;

    char dateString[] = "1975-11-10T01:25:00";
    tm parsedDate = parseDate((dateString));
    char formattedParsedDate[50];
    strftime(formattedParsedDate, 50, "%FT%T", &parsedDate);
    cout << "Parsed date is " << formattedParsedDate;


    return 0;
}